ledger/complete/mtrie/trie/trie.go

```diff
@@ -324,13 +363,22 @@ func update(
 				// the case where the recursion stops: only one path to update
 				if len(paths) == 1 {
 					if !parentNode.Payload().Equals(&payloads[i]) {
```

Comment:
@tarakby this is the way we use _using_ the trie. But note that the trie itself does not do the hashing from `key` to path. From the perspective of the Trie, the `Key` is auxiliary information that the higher level code (outside of the MTrie's control), can set and change at will. 
In your argument, you are essentially proposing to make implicit assumptions in the lower level code about how it will be used. Even if these assumptions are true right now, the Mtrie cannot enforce them. And without detailed knowledge of the MTrie an engineer will not know that they potentially break these assumptions when working on the higher-level code. This makes the code brittle. 

Personally, I prefer if the MTrie would handle _every_ usage pattern, including the case where the `Key` might change arbitrarily. But I don't think this is possible for this code without significant changes. Therefore, I suggested that the API clearly documents the MTrie's assumption, which is the second best option.

Reply:
Yes that's right. We shouldn't make such implicit assumptions. I've seen you mentioned a similar argument in this PR too 👌🏼